### PR TITLE
[14.0][MIG] sale exception

### PR DESCRIPTION
**Pull request by damdam-s**
superseeds #1389

@Saran440 @atchuthan @sebastienbeau @gurneyalex

### CODE REVIEW

**Review by atchuthan — APPROVED** [review]

File: sale_exception/models/sale.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    _inherit = ["sale.order", "base.exception"]
+    _name = "sale.order"
+    _order = "main_exception_id asc, date_order desc, name desc"
+
+    @api.model
+    def _reverse_field(self):
+        return "sale_ids"
+
+    def detect_exceptions(self):
+        all_exceptions = super(SaleOrder, self).detect_exceptions()
```
Comment (jcoux) [inline sale_exception/models/sale.py]:
```suggestion
        all_exceptions = super().detect_exceptions()
```

File: sale_exception/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        order_set = self.search([("state", "=", "draft")])
+        order_set.detect_exceptions()
+        return True
+
+    def _fields_trigger_check_exception(self):
+        return ["ignore_exception", "order_line", "state"]
+
+    @api.model
+    def create(self, vals):
+        record = super(SaleOrder, self).create(vals)
```
Comment (jcoux) [inline sale_exception/models/sale.py]:
```suggestion
        record = super().create(vals)
```

File: sale_exception/models/sale.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        check_exceptions = any(
+            field in vals for field in self._fields_trigger_check_exception()
+        )
+        if check_exceptions:
+            record.sale_check_exception()
+        return record
+
+    def write(self, vals):
+        result = super(SaleOrder, self).write(vals)
```
Comment (jcoux) [inline sale_exception/models/sale.py]:
```suggestion
        result = super().write(vals)
```

File: sale_exception/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        check_exceptions = any(
+            field in vals for field in self._fields_trigger_check_exception()
+        )
+        if check_exceptions:
+            self.sale_check_exception()
```
Comment (jcoux) [inline sale_exception/models/sale.py]: Would have these lines in a dedicated function as we have the same check on create and write function.

File: sale_exception/models/sale.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            self.ignore_exception = False
+
+    def action_confirm(self):
+        if self.detect_exceptions():
+            return self._popup_exceptions()
+        return super().action_confirm()
+
+    def action_draft(self):
+        res = super().action_draft()
+        orders = self.filtered(lambda s: s.ignore_exception)
```
Comment (jcoux) [inline sale_exception/models/sale.py]:
```suggestion
        orders = self.filtered("ignore_exception")
```

**Review by jcoux — CHANGES_REQUESTED** [review]

**Review by jcoux — APPROVED** [review]
Approved (if travis/runbot is green)

**Review by rousseldenis — APPROVED** [review]

**Comment by rousseldenis** [comment]
@damdam-s Could you squash ?

**Comment by damdam-s** [comment]
done

**Pull request merged**